docs/manual/resource-leak-checker.tex

```diff
@@ -588,6 +597,234 @@
+
+\begin{verbatim}
+@OwningArray Socket[] s1 = new Socket[n];
+@OwningArray Socket[] s2 = new Socket[n];
+s2 = s1; // prohibited assignment
+\end{verbatim}
+
+In general is aliasing of any kind for \<@OwningArray> variables not permitted. In particular, a non-\<@OwningArray> variable may also not be assigned to an \<@OwningArray> variable. This implies that an \<@OwningArray> variable is always the only reference to the underlying array within a procedure. The exception occurs when an \<@OwningArray> referenc is passed as an argument. More on that later.
+
+\subsectionAndLabel{Must-call and called-methods for arrays}{resource-leak-arrays-mcoe-cmoe}
+Must-call-on-elements and called-methods-on-elements provide similar semantics for arrays as must-call and called-methods do for objects.
+
+The type hierarchy is the following:
+
+\begin{verbatim}
+              MustCallOnElementsUnknown
+                           |
+            MustCallOnElements({"a", "b"})
+               /                     \
+MustCallOnElements({"b"})       MustCallOnElements({"a"})
+               \                    /
+               MustCallOnElements({})
+\end{verbatim}
+The default type is \<@MustCallOnElements({})> (bottom) for everything except \<@OwningArray> fields and \<@OwningArray> parameters. For those, the default \<@MustCallOnElements> value is the \<@MustCall> value of the component type. For example, the default type of an \<@OwningArray> Socket[] is \<@MustCallOnElements({``close''})>, since Socket has \<@MustCall(``close'')> type.
+
+The fulfilled obligations are stored in the \<@CalledMethodsOnElements> type annotation with the following type hierarchy:
+\begin{verbatim}
+                      CalledMethodsOnElements({})
+                      /                        \
+CalledMethodsOnElements({"b"})           CalledMethodsOnElements({"a"})
+                      \                        /
+                 CalledMethodsOnElements({"a", "b"})
+                                   |
+                    CalledMethodsOnElementsBottom
+\end{verbatim}
+The default type is \<@CalledMethodsOnElements({})> (top) for any declaration.
+
+\subsectionAndLabel{Directly fulfilling calling obligations on \<@OwningArray> arrays}{resource-leak-array-fulfilling-obligations}
+
+The MustCallOnElements obligations can be fulfilled in the following way:
+\begin{verbatim}
+// 1. @MustCallOnElements({"close"}) @CalledMethodsOnElements({})
+for (int i = 0; i < arr.length; i++) {
+    arr[i].close();
+}
+// 2. @MustCallOnElements({}) @CalledMethodsOnElements({"close"})
+\end{verbatim}
+
+The same pattern-match rules as for the assignment of elements apply (see Section~\ref{resource-leak-arrays-declaration}).
```

Comment:
I feel like a `try-catch` statement in the loop body should be required, here?  Otherwise, if `close()` throws an exception for an element early in the array, and that exception causes control flow to break out of the loop, `close()` is never called on the remaining array elements.

Reply:
maybe? it's not specified what type the elements of the array have. If it's Socket, then yes. If it's something that doesn't explicitly throw an exception when constructed, do we still need it? Technically, any java instruction could throw an exception, right?

I added a try-catch anyways for clarity.